### PR TITLE
Fix context.authentication undefined

### LINKED ISSUE #267: Context.Authentication possibly undefinded
https://github.com/auth0/rules/blob/dc917b5d7cce9256243abbeffd2f4725b502c4c1/src/rules/require-mfa-once-per-session.js#L14

In the last few days, with no other changes to our auth0 config, we noticed that in this rule, context.authentication is sometimes undefined, causing an error to throw on this rule, unauthenticating our users temporarily.  Should there be a scenario where this is possible?

### PR DESCRIPTION
**Pull request by joshcanhelp**
Fixes #267

### CODE REVIEW

File: test/rules/guardian-multifactor-stepup-authentication.test.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const loadRule = require('../utils/load-rule');
 const ContextBuilder = require('../utils/contextBuilder');
 const RequestBuilder = require('../utils/requestBuilder');
 
-const ruleName = 'guardian-multifactor-ip-range';
+const ruleName = 'guardian-multifactor-stepup-authentication';
```
Comment (joshcanhelp) [inline test/rules/guardian-multifactor-stepup-authentication.test.js:7]: Was not testing the right Rule

File: test/rules/guardian-multifactor-stepup-authentication.test.js [hunk 2]
```diff
@@ -15,34 +15,66 @@ describe(ruleName, () => {
...
+    it('does not set the multifactor context', (done) => {
+      rule(user, context, (err, u, c) => {
+        expect(c.multifactor).toBeFalsy();
+
+        done();
+      });
+    });
+  });
+
+  describe('context.authentication includes mfa', () => {
```
Comment (joshcanhelp) [inline test/rules/guardian-multifactor-stepup-authentication.test.js:61]: Test for the fix

File: test/utils/requestBuilder.js [hunk 3]
```diff
@@ -38,6 +38,10 @@ class RequestBuilder {
     this.request.geoip = geoIp;
     return this;
   }
+  withQuery(query) {
+    this.request.query = { ...this.request.query, ...query };
```
Comment (joshcanhelp) [inline test/utils/requestBuilder.js:42]: `query` should be shallow

File: src/rules/guardian-multifactor-stepup-authentication.js [hunk 4]
```diff
@@ -21,11 +21,16 @@ function guardianMultifactorStepUpAuthentication(user, context, callback) {
...
-  if (
+  const isMfa =
     context.request.query.acr_values ===
-      'http://schemas.openid.net/pape/policies/2007/06/multi-factor' &&
-    !context.authentication.methods.some((method) => method.name === 'mfa')
-  ) {
+    'http://schemas.openid.net/pape/policies/2007/06/multi-factor';
+
+  let authMethods = [];
+  if (context.authentication && Array.isArray(context.authentication.methods)) {
```
Comment (joshcanhelp) [inline src/rules/guardian-multifactor-stepup-authentication.js:29]: Fix here

File: src/rules/require-mfa-once-per-session.js [hunk 5]
```diff
@@ -11,9 +11,12 @@
  */
 
 function requireMfaOncePerSession(user, context, callback) {
-  const completedMfa = !!context.authentication.methods.find(
-    (method) => method.name === 'mfa'
-  );
+  let authMethods = [];
+  if (context.authentication && Array.isArray(context.authentication.methods)) {
```
Comment (joshcanhelp) [inline src/rules/require-mfa-once-per-session.js:15]: Same problem here

**Review by shawnmclean — APPROVED** [review]

**Pull request merged**